core/src/main/java/io/questdb/cairo/mv/DefinitionFileWriter.java

```diff
@@ -0,0 +1,428 @@
+import io.questdb.cairo.vm.Vm;
+import io.questdb.cairo.vm.api.MemoryCARW;
+import io.questdb.cairo.vm.api.MemoryCMARW;
+import io.questdb.std.BinarySequence;
+import io.questdb.std.FilesFacade;
+import io.questdb.std.Long256;
+import io.questdb.std.MemoryTag;
+import io.questdb.std.Misc;
+import io.questdb.std.Mutable;
+import io.questdb.std.Transient;
+import io.questdb.std.Unsafe;
+import io.questdb.std.Vect;
+import io.questdb.std.str.LPSZ;
+import io.questdb.std.str.Utf8Sequence;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+import static io.questdb.cairo.mv.DefinitionFileUtils.*;
+
+public class DefinitionFileWriter implements Closeable, Mutable {
+    private final BlockMemoryHandleImpl blockMemoryHandle = new BlockMemoryHandleImpl();
+    private final FilesFacade ff;
+    private int blockCount;
+    private long blockOffset;
+    private MemoryCMARW file;
+    private boolean isCommitted;
+    private MemoryCARW memory;
+
+    public DefinitionFileWriter(FilesFacade ff) {
+        this.ff = ff;
+    }
+
+    public AppendOnlyBlock append() {
+        return blockMemoryHandle.reset(0);
+    }
+
+    @Override
+    public void clear() {
+        Misc.free(memory);
+        Misc.free(file);
+        isCommitted = false;
+        blockOffset = 0;
+        blockCount = 0;
+    }
+
+    @Override
+    public void close() throws IOException {
+        clear();
+        memory = null;
+        file = null;
+    }
+
+    public boolean commit() {
+        if (isCommitted) {
+            return false;
+        }
+
+        // exclude checksum but include block count
+        final long memoryBaseAddress = memory.getPageAddress(0);
+        final long checksumAddress = memoryBaseAddress + REGION_HEADER_SIZE + REGION_BLOCK_COUNT_OFFSET;
+        final long checksumSize = blockOffset - REGION_HEADER_SIZE - REGION_BLOCK_COUNT_OFFSET;
+        final int checksum = getChecksum(checksumAddress, checksumSize);
+
+        memory.putInt(REGION_CHECKSUM_OFFSET, checksum);
+        memory.putInt(REGION_BLOCK_COUNT_OFFSET, blockCount);
+
+        long currentVersion = getVersionVolatile();
+        final long regionLength = blockOffset;
+        final long currentRegionOffset = getRegionOffset(currentVersion);
+        final long regionWriteOffset = (regionLength <= currentRegionOffset) ?
+                0 : currentRegionOffset + getRegionLength(currentVersion);
+
+        file.jumpTo(regionWriteOffset + HEADER_SIZE + regionLength);
+        final long fileBaseAddress = file.getPageAddress(0);
+        Vect.memcpy(fileBaseAddress + regionWriteOffset + HEADER_SIZE, memoryBaseAddress, regionLength);
+
+        currentVersion += 1;
+        setRegionOffset(currentVersion, regionWriteOffset);
+        setRegionLength(currentVersion, regionLength);
+        setVersionVolatile(currentVersion);
+
+        isCommitted = true;
+        file.sync(false);
```

Comment:
Shouldn't we check commit mode here and only sync when it's not nosync?

Reply:
addressed